{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "72883641-314e-496c-a41e-1170f0ca8e71",
   "metadata": {},
   "source": [
    "# Rsult analysis for differential gene expression for samples from BeatAML study*\n",
    "\n",
    "This is a continuation of differential gene expression (DGE) performed for samples of the study by Tyner et al. (2018).\n",
    "\n",
    "This is a very basic functional analysis performed for genes differentially expressed between the compared groups (samples from patients with or without a mutation in NPM1 gene). Besides DGE results from the previous step, a dataset linking gene IDs to GO annotations is required.\n",
    "\n",
    "The data on GO annotaions is already saved in the input directory as _`mart_export_go.tsv.gz`_. It can be redownloaded in the following way:\n",
    "- go to https://www.ensembl.org/biomart/martview\n",
    "- as `-CHOOSE DATABASE-` select `Ensembl Genes 109`\n",
    "- as `-CHOOSE DATASET-` select `Human Genes (GRCh38.p13)`\n",
    "- go to `Attributes` and in the section `GENES:` select:\n",
    "    - Gene stable ID\n",
    "- and in the section `EXTERNALS:` (subsection _Go_) select:\n",
    "    - GO term accession\n",
    "    - GO term name**\n",
    "    - GO term evidence code\n",
    "    - GO domain\n",
    "- unselect other then listed above attributes in all sections if such were preselected\n",
    "- click the `Results` button\n",
    "- in `Export all results to` choose `Compressed file (.gz)`, `TSV`\n",
    "- check the option `Unique results only`\n",
    "- click the `Go` button\n",
    "- save the file in a location of your choosing\n",
    "\n",
    "The logic behind the analysis is concisely explained step by step.\n",
    "\n",
    "---\n",
    "\n",
    "<font size=\"1\">\n",
    "*Tyner JW, Tognon CE, Bottomly D, Wilmot B, Kurtz SE, Savage SL, Long N, Schultz AR, Traer E, Abel M, Agarwal A. Functional genomic landscape of acute myeloid leukaemia. Nature. 2018 Oct 25;562(7728):526-31.\n",
    "    \n",
    "**This is actually the only column that will be utilised, however, it is good to have others at hand (just in case).\n",
    "</font>\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb753ac8-1c41-4303-8b4a-231de5843b5e",
   "metadata": {},
   "source": [
    "- Import Pandas library and load DGE results to `res_df` DataFrame. Look up the whole DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "671642dc-42a6-46b0-8474-1fccc106949b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fcac7193-f262-42ff-967e-75c9bcabd0d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>21.614238</td>\n",
       "      <td>0.159750</td>\n",
       "      <td>0.211504</td>\n",
       "      <td>0.364023</td>\n",
       "      <td>0.518267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>0.117638</td>\n",
       "      <td>-0.107868</td>\n",
       "      <td>0.373021</td>\n",
       "      <td>0.724421</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>1122.078004</td>\n",
       "      <td>-0.100693</td>\n",
       "      <td>0.043725</td>\n",
       "      <td>0.020318</td>\n",
       "      <td>0.057731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>685.212410</td>\n",
       "      <td>0.181222</td>\n",
       "      <td>0.061381</td>\n",
       "      <td>0.002753</td>\n",
       "      <td>0.011163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>390.516247</td>\n",
       "      <td>-0.183616</td>\n",
       "      <td>0.077746</td>\n",
       "      <td>0.015718</td>\n",
       "      <td>0.047020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63672</th>\n",
       "      <td>ENSG00000273480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63673</th>\n",
       "      <td>ENSG00000273482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63674</th>\n",
       "      <td>ENSG00000273484</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63675</th>\n",
       "      <td>ENSG00000273490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63676</th>\n",
       "      <td>ENSG00000273491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63677 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id     baseMean  log2FoldChange     lfcSE    pvalue  \\\n",
       "0      ENSG00000000003    21.614238        0.159750  0.211504  0.364023   \n",
       "1      ENSG00000000005     0.117638       -0.107868  0.373021  0.724421   \n",
       "2      ENSG00000000419  1122.078004       -0.100693  0.043725  0.020318   \n",
       "3      ENSG00000000457   685.212410        0.181222  0.061381  0.002753   \n",
       "4      ENSG00000000460   390.516247       -0.183616  0.077746  0.015718   \n",
       "...                ...          ...             ...       ...       ...   \n",
       "63672  ENSG00000273480          NaN             NaN       NaN       NaN   \n",
       "63673  ENSG00000273482          NaN             NaN       NaN       NaN   \n",
       "63674  ENSG00000273484          NaN             NaN       NaN       NaN   \n",
       "63675  ENSG00000273490          NaN             NaN       NaN       NaN   \n",
       "63676  ENSG00000273491          NaN             NaN       NaN       NaN   \n",
       "\n",
       "           padj  \n",
       "0      0.518267  \n",
       "1           NaN  \n",
       "2      0.057731  \n",
       "3      0.011163  \n",
       "4      0.047020  \n",
       "...         ...  \n",
       "63672       NaN  \n",
       "63673       NaN  \n",
       "63674       NaN  \n",
       "63675       NaN  \n",
       "63676       NaN  \n",
       "\n",
       "[63677 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df = pd.read_csv('output/DGE_results.tsv', sep='\\t')\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad70a52f-5da9-4e75-8b8e-917d0df53fa3",
   "metadata": {},
   "source": [
    "- Filter `res_df` accroding to `log2FoldChange` of at least `1.0` (equivalent of fold change `2.0`, both in plus and in minus directions) and `FDR` as the one indicated during performing DGE analysis (`0.01`). Keep the results in `fin_df` DataFrame. Look up the whole resulting DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ab20d5fc-b162-4bb7-bdf3-42bb6c5e67a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000224842</td>\n",
       "      <td>2.828884</td>\n",
       "      <td>5.220453</td>\n",
       "      <td>0.476137</td>\n",
       "      <td>2.086320e-28</td>\n",
       "      <td>6.814681e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>1.084612</td>\n",
       "      <td>4.639766</td>\n",
       "      <td>0.503022</td>\n",
       "      <td>6.618996e-06</td>\n",
       "      <td>6.072014e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000133110</td>\n",
       "      <td>7.404256</td>\n",
       "      <td>4.627742</td>\n",
       "      <td>0.460003</td>\n",
       "      <td>5.028314e-10</td>\n",
       "      <td>1.164092e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000018236</td>\n",
       "      <td>118.647381</td>\n",
       "      <td>4.566301</td>\n",
       "      <td>0.382213</td>\n",
       "      <td>2.329092e-33</td>\n",
       "      <td>1.321331e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000136944</td>\n",
       "      <td>32.970138</td>\n",
       "      <td>4.407800</td>\n",
       "      <td>0.407030</td>\n",
       "      <td>1.858105e-44</td>\n",
       "      <td>3.338086e-41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2722</th>\n",
       "      <td>ENSG00000213931</td>\n",
       "      <td>78.940749</td>\n",
       "      <td>-6.978391</td>\n",
       "      <td>0.388769</td>\n",
       "      <td>6.072773e-11</td>\n",
       "      <td>1.716567e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2723</th>\n",
       "      <td>ENSG00000266976</td>\n",
       "      <td>1.882522</td>\n",
       "      <td>-7.004681</td>\n",
       "      <td>0.648726</td>\n",
       "      <td>3.675058e-07</td>\n",
       "      <td>4.624664e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2724</th>\n",
       "      <td>ENSG00000206579</td>\n",
       "      <td>1.685687</td>\n",
       "      <td>-7.073905</td>\n",
       "      <td>1.190329</td>\n",
       "      <td>2.160649e-05</td>\n",
       "      <td>1.729039e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2725</th>\n",
       "      <td>ENSG00000147255</td>\n",
       "      <td>13.753163</td>\n",
       "      <td>-7.236593</td>\n",
       "      <td>0.699507</td>\n",
       "      <td>7.802278e-21</td>\n",
       "      <td>1.042571e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2726</th>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>9.432025</td>\n",
       "      <td>-8.061118</td>\n",
       "      <td>0.528245</td>\n",
       "      <td>4.472088e-17</td>\n",
       "      <td>3.355311e-15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2727 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id    baseMean  log2FoldChange     lfcSE        pvalue  \\\n",
       "0     ENSG00000224842    2.828884        5.220453  0.476137  2.086320e-28   \n",
       "1     ENSG00000107807    1.084612        4.639766  0.503022  6.618996e-06   \n",
       "2     ENSG00000133110    7.404256        4.627742  0.460003  5.028314e-10   \n",
       "3     ENSG00000018236  118.647381        4.566301  0.382213  2.329092e-33   \n",
       "4     ENSG00000136944   32.970138        4.407800  0.407030  1.858105e-44   \n",
       "...               ...         ...             ...       ...           ...   \n",
       "2722  ENSG00000213931   78.940749       -6.978391  0.388769  6.072773e-11   \n",
       "2723  ENSG00000266976    1.882522       -7.004681  0.648726  3.675058e-07   \n",
       "2724  ENSG00000206579    1.685687       -7.073905  1.190329  2.160649e-05   \n",
       "2725  ENSG00000147255   13.753163       -7.236593  0.699507  7.802278e-21   \n",
       "2726  ENSG00000079102    9.432025       -8.061118  0.528245  4.472088e-17   \n",
       "\n",
       "              padj  \n",
       "0     6.814681e-26  \n",
       "1     6.072014e-05  \n",
       "2     1.164092e-08  \n",
       "3     1.321331e-30  \n",
       "4     3.338086e-41  \n",
       "...            ...  \n",
       "2722  1.716567e-09  \n",
       "2723  4.624664e-06  \n",
       "2724  1.729039e-04  \n",
       "2725  1.042571e-18  \n",
       "2726  3.355311e-15  \n",
       "\n",
       "[2727 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_df = res_df[\n",
    "    (res_df['log2FoldChange'].abs() >= 1.0 ) & (res_df['padj'] <= 0.01)\n",
    "].sort_values('log2FoldChange', ascending=False).reset_index(drop=True)\n",
    "fin_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1be613de-2916-425e-8092-95170eb9a82e",
   "metadata": {},
   "source": [
    "- Load to `meta_df` DataFrame the dataset obtained from Ensembl BioMart that links gene IDs with GO annotations. Look up the whole DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d7451605-561e-432c-af52-e93cf5a1978e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>GO term accession</th>\n",
       "      <th>GO term name</th>\n",
       "      <th>GO term evidence code</th>\n",
       "      <th>GO domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>GO:0030533</td>\n",
       "      <td>triplet codon-amino acid adaptor activity</td>\n",
       "      <td>IEA</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>GO:0006412</td>\n",
       "      <td>translation</td>\n",
       "      <td>IEA</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>GO:0003735</td>\n",
       "      <td>structural constituent of ribosome</td>\n",
       "      <td>IEA</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>GO:0005840</td>\n",
       "      <td>ribosome</td>\n",
       "      <td>IEA</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601467</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>GO:0001669</td>\n",
       "      <td>acrosomal vesicle</td>\n",
       "      <td>IDA</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601468</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>GO:0031410</td>\n",
       "      <td>cytoplasmic vesicle</td>\n",
       "      <td>IDA</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601469</th>\n",
       "      <td>ENSG00000284882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601470</th>\n",
       "      <td>ENSG00000289881</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601471</th>\n",
       "      <td>ENSG00000285325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601472 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene stable ID GO term accession  \\\n",
       "0       ENSG00000210049        GO:0030533   \n",
       "1       ENSG00000210049        GO:0006412   \n",
       "2       ENSG00000211459        GO:0003735   \n",
       "3       ENSG00000211459        GO:0005840   \n",
       "4       ENSG00000210077               NaN   \n",
       "...                 ...               ...   \n",
       "601467  ENSG00000122432        GO:0001669   \n",
       "601468  ENSG00000122432        GO:0031410   \n",
       "601469  ENSG00000284882               NaN   \n",
       "601470  ENSG00000289881               NaN   \n",
       "601471  ENSG00000285325               NaN   \n",
       "\n",
       "                                     GO term name GO term evidence code  \\\n",
       "0       triplet codon-amino acid adaptor activity                   IEA   \n",
       "1                                     translation                   IEA   \n",
       "2              structural constituent of ribosome                   IEA   \n",
       "3                                        ribosome                   IEA   \n",
       "4                                             NaN                   NaN   \n",
       "...                                           ...                   ...   \n",
       "601467                          acrosomal vesicle                   IDA   \n",
       "601468                        cytoplasmic vesicle                   IDA   \n",
       "601469                                        NaN                   NaN   \n",
       "601470                                        NaN                   NaN   \n",
       "601471                                        NaN                   NaN   \n",
       "\n",
       "                 GO domain  \n",
       "0       molecular_function  \n",
       "1       biological_process  \n",
       "2       molecular_function  \n",
       "3       cellular_component  \n",
       "4                      NaN  \n",
       "...                    ...  \n",
       "601467  cellular_component  \n",
       "601468  cellular_component  \n",
       "601469                 NaN  \n",
       "601470                 NaN  \n",
       "601471                 NaN  \n",
       "\n",
       "[601472 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df = pd.read_csv('input/mart_export_go.tsv.gz', sep='\\t')\n",
    "meta_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6fce61c-6257-49de-a347-e91ad5a60bf7",
   "metadata": {},
   "source": [
    "- Merge both `fin_df` with `meta_df` on gene_id/Gene stable ID columns using only gene IDs from `fin_df` in order to obtain `go_df` that links gene IDs for differientiating transcripts with GO annotations. Look up the whole resulting DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b3fe5a06-c099-436f-a211-9c76083be516",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>GO term accession</th>\n",
       "      <th>GO term name</th>\n",
       "      <th>GO term evidence code</th>\n",
       "      <th>GO domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>GO:0003677</td>\n",
       "      <td>DNA binding</td>\n",
       "      <td>IEA</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>regulation of DNA-templated transcription</td>\n",
       "      <td>IEA</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>GO:0003700</td>\n",
       "      <td>DNA-binding transcription factor activity</td>\n",
       "      <td>IEA</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>GO:0000981</td>\n",
       "      <td>DNA-binding transcription factor activity, RNA...</td>\n",
       "      <td>IEA</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>ENSG00000107807</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>IEA</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50893</th>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>GO:0045892</td>\n",
       "      <td>negative regulation of DNA-templated transcrip...</td>\n",
       "      <td>IDA</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50894</th>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>GO:0045599</td>\n",
       "      <td>negative regulation of fat cell differentiation</td>\n",
       "      <td>ISS</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50895</th>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>GO:0003714</td>\n",
       "      <td>transcription corepressor activity</td>\n",
       "      <td>TAS</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50896</th>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>GO:0003714</td>\n",
       "      <td>transcription corepressor activity</td>\n",
       "      <td>IDA</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50897</th>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>ENSG00000079102</td>\n",
       "      <td>GO:0140297</td>\n",
       "      <td>DNA-binding transcription factor binding</td>\n",
       "      <td>IPI</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48696 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id   Gene stable ID GO term accession  \\\n",
       "1      ENSG00000107807  ENSG00000107807        GO:0003677   \n",
       "2      ENSG00000107807  ENSG00000107807        GO:0006355   \n",
       "3      ENSG00000107807  ENSG00000107807        GO:0003700   \n",
       "4      ENSG00000107807  ENSG00000107807        GO:0000981   \n",
       "5      ENSG00000107807  ENSG00000107807        GO:0005634   \n",
       "...                ...              ...               ...   \n",
       "50893  ENSG00000079102  ENSG00000079102        GO:0045892   \n",
       "50894  ENSG00000079102  ENSG00000079102        GO:0045599   \n",
       "50895  ENSG00000079102  ENSG00000079102        GO:0003714   \n",
       "50896  ENSG00000079102  ENSG00000079102        GO:0003714   \n",
       "50897  ENSG00000079102  ENSG00000079102        GO:0140297   \n",
       "\n",
       "                                            GO term name  \\\n",
       "1                                            DNA binding   \n",
       "2              regulation of DNA-templated transcription   \n",
       "3              DNA-binding transcription factor activity   \n",
       "4      DNA-binding transcription factor activity, RNA...   \n",
       "5                                                nucleus   \n",
       "...                                                  ...   \n",
       "50893  negative regulation of DNA-templated transcrip...   \n",
       "50894    negative regulation of fat cell differentiation   \n",
       "50895                 transcription corepressor activity   \n",
       "50896                 transcription corepressor activity   \n",
       "50897           DNA-binding transcription factor binding   \n",
       "\n",
       "      GO term evidence code           GO domain  \n",
       "1                       IEA  molecular_function  \n",
       "2                       IEA  biological_process  \n",
       "3                       IEA  molecular_function  \n",
       "4                       IEA  molecular_function  \n",
       "5                       IEA  cellular_component  \n",
       "...                     ...                 ...  \n",
       "50893                   IDA  biological_process  \n",
       "50894                   ISS  biological_process  \n",
       "50895                   TAS  molecular_function  \n",
       "50896                   IDA  molecular_function  \n",
       "50897                   IPI  molecular_function  \n",
       "\n",
       "[48696 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_df = fin_df[['gene_id']].merge(meta_df, how='left', left_on='gene_id', right_on='Gene stable ID')\n",
    "go_df = go_df[ go_df['GO term name'].notna() ]\n",
    "go_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e84c52ca-9646-4869-b665-3be79f06392b",
   "metadata": {},
   "source": [
    "- Perform a very basic analysis by calculating % of differientiating genes falling into GO terms. Since one gene may be described with more than one GO term, the resulting values will obviously not sum to 100%. Display GO terms that at lest 5% genes fell into."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7348d7a6-f45d-4e81-987e-4d3870ddb353",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO term name</th>\n",
       "      <th>% of genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>62.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>protein binding</td>\n",
       "      <td>50.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>membrane</td>\n",
       "      <td>42.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>29.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>extracellular region</td>\n",
       "      <td>28.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>25.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>extracellular space</td>\n",
       "      <td>21.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cytosol</td>\n",
       "      <td>17.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>metal ion binding</td>\n",
       "      <td>12.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>signal transduction</td>\n",
       "      <td>10.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DNA-binding transcription factor activity, RNA...</td>\n",
       "      <td>9.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nucleoplasm</td>\n",
       "      <td>9.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>extracellular exosome</td>\n",
       "      <td>8.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DNA binding</td>\n",
       "      <td>7.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>7.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cell adhesion</td>\n",
       "      <td>7.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>integral component of membrane</td>\n",
       "      <td>7.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>positive regulation of transcription by RNA po...</td>\n",
       "      <td>6.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Golgi apparatus</td>\n",
       "      <td>6.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>cell surface</td>\n",
       "      <td>6.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>synapse</td>\n",
       "      <td>6.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>regulation of transcription by RNA polymerase II</td>\n",
       "      <td>6.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>identical protein binding</td>\n",
       "      <td>6.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>calcium ion binding</td>\n",
       "      <td>6.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>RNA polymerase II cis-regulatory region sequen...</td>\n",
       "      <td>6.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>cell differentiation</td>\n",
       "      <td>5.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>immune response</td>\n",
       "      <td>5.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>G protein-coupled receptor signaling pathway</td>\n",
       "      <td>5.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>collagen-containing extracellular matrix</td>\n",
       "      <td>5.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>regulation of DNA-templated transcription</td>\n",
       "      <td>5.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>immune system process</td>\n",
       "      <td>5.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>cell projection</td>\n",
       "      <td>5.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>nervous system development</td>\n",
       "      <td>5.02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         GO term name  % of genes\n",
       "0                                     plasma membrane       62.56\n",
       "1                                     protein binding       50.79\n",
       "2                                            membrane       42.46\n",
       "3                                           cytoplasm       29.48\n",
       "4                                extracellular region       28.97\n",
       "5                                             nucleus       25.93\n",
       "6                                 extracellular space       21.20\n",
       "7                                             cytosol       17.68\n",
       "8                                   metal ion binding       12.10\n",
       "9                                 signal transduction       10.63\n",
       "10  DNA-binding transcription factor activity, RNA...        9.72\n",
       "11                                        nucleoplasm        9.06\n",
       "12                              extracellular exosome        8.76\n",
       "13                                        DNA binding        7.96\n",
       "14                              endoplasmic reticulum        7.59\n",
       "15                                      cell adhesion        7.37\n",
       "16                     integral component of membrane        7.00\n",
       "17  positive regulation of transcription by RNA po...        6.93\n",
       "18                                    Golgi apparatus        6.78\n",
       "19                                       cell surface        6.45\n",
       "20                                            synapse        6.45\n",
       "21   regulation of transcription by RNA polymerase II        6.38\n",
       "22                          identical protein binding        6.23\n",
       "23                                calcium ion binding        6.23\n",
       "24  RNA polymerase II cis-regulatory region sequen...        6.16\n",
       "25                               cell differentiation        5.90\n",
       "26                                    immune response        5.90\n",
       "27       G protein-coupled receptor signaling pathway        5.87\n",
       "28           collagen-containing extracellular matrix        5.68\n",
       "29          regulation of DNA-templated transcription        5.57\n",
       "30                              immune system process        5.39\n",
       "31                                    cell projection        5.24\n",
       "32                         nervous system development        5.02"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_s = (go_df.value_counts('GO term name')/fin_df.shape[0]*100).round(2)\n",
    "counts_s = counts_s[counts_s >= 5.0]\n",
    "counts_s.name = '% of genes'\n",
    "counts_df = counts_s.to_frame().reset_index()\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7ef6d02-43dc-41cc-976c-c827eaa8107f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
